DietPi-Banner | Word wrap for small terminals

mtekman opened the pull request:
Dietpi-Banner word wrap using MAWK

This is a 2025 re-implementation of https://github.com/MichaIng/DietPi/pull/5820, based on comments from https://github.com/MichaIng/DietPi/pull/5820#issuecomment-3395425672

Commits:
- Print_Header has largely been indented, please ignore 0b05ae63fff8d8008c1499ccc4cb698d2dffa5ee for a cleaner PR
- Bannerwrap opts added to a separate file, updated by G_WHIP_MENU_ARRAY for option 19

mtekman committed:
Bannerwrap initial awk and opts files

mtekman committed:
whitespace indent -- skip this commit if needed

mtekman committed:
fold command added to dietpi-banner with cat passthrough when "none"

mtekman committed:
Custom Banner entry converted into an if statement

mtekman committed:
Add bannerwrap interactive options

mtekman commented on dietpi/func/dietpi-banner:
```diff
@@ -245,65 +248,77 @@ $GREEN_LINE"
+			Print_Updates
+			(( ${aENABLED[11]} )) && Print_Useful_Commands
+		) | ${FOLD_CMD}
+		echo "fold_cmd=${FOLD_CMD}"
```
Use for debugging, please remove this line after testing/verifying that all seems correct

mtekman committed:
added shebang

mtekman commented on dietpi/misc/bannerwrap_fold.awk:
```diff
@@ -0,0 +1,99 @@
+
+function make_whitespace_offset(clean_line){
+    ## Generate leading spaces based on global INDENT_TYPE and
+    ## a tokenized list of words.
```
```suggestion
    ## Generate leading spaces based on global INDENT_TYPE
```

mtekman committed:
Apply suggestion from @mtekman

MichaIng committed:
dietpi-banner: code alignments

* Remove BW_MAXCOL variable, but obtain terminal width on demand
* Declare word-wrap defaults in main script and store to main config, remove dedicated word-wrap config
* Remove word-warp option "none", instead use word-wrapping based on main menu toggle "${aENABLED[19]}"
* Turn FOLD_CMD variable into an array for failsafe argument splitting
* Call banner print commands as command group rather than as expensive subshell
* Remove debug output

MichaIng committed:
dietpi-banner: save settings once after processing whole options loop

mtekman commented on dietpi/func/dietpi-banner-wrap.awk:
```diff
@@ -0,0 +1,98 @@
+    wn = split($0, words, /[[:space:]]+/)
+    sn = split(clean_line, stripped_words, /[[:space:]]+/)
+
+    if (sn > 0 && wn != sn){
+        print "[ERROR] Word ordering does not match stripped word ordering"
+        exit 1
+    }
```
Maybe this should fail with a warning or silently, otherwise I wonder if anything will be printed at all

MichaIng replied:
As long as the error is printed to console it should be fine. I'll check how this condition can happen at all. In case the error should be meaningful and link our bug tracker? I assume it is not something that happens expectedly by times.

mtekman replied:
I have never seen the error trigger on the control/color codes I've tested it with, but I can envision a line that might have garbled text for one reason or another, that when tokenized into words does not match with the number of tokenized color words.

Worst case scenario is that the line is not folded properly. I think ideally it should be more of a warning than something that prevents the rest of the lines from being processed

mtekman commented on dietpi/misc/bannerwrap_fold.awk:
```diff
@@ -0,0 +1,98 @@
+
+    if (sn > 0 && wn != sn){
+        print "[ERROR] Word ordering does not match stripped word ordering"
+        exit 1
```
```suggestion
        print "[WARN] Color word order does not match stripped word order"
        next
```

MichaIng replied:
Okay, makes sense.

MichaIng committed:
dietpi-banner: fix MAXCOL variable and skip cat pipe

Skip call and pipe to external cat command it word-wrap is not enabled by splitting the unwrapped output into a dedicated function, and pipe the output of that function into mawk if enabled.

Fix MAXCOL variable name in awk script, and turn word order error into warning.

MichaIng committed:
dietpi-banner: fix additional INDENT_MIN space

and rename awk script to make clear that it is part of the dietpi-banner.

MichaIng committed:
dietpi-banner: fix leading spaces for explicit "min" option

If INDENT_MIN is used, either since still too long after dash/colon alignment, if if "min" was explicitly selected, do not add an extra space. Adjust leading space correction for dash/colon to keep them correctly aligned.

MichaIng commented on dietpi/func/dietpi-banner:
```diff
@@ -327,19 +350,35 @@ $GREEN_LINE"
+				)
+				G_WHIP_DEFAULT_ITEM=$BW_INDENT_TYPE
+				G_WHIP_MENU 'Choose the type of word-wrapping' && BW_INDENT_TYPE=$G_WHIP_RETURNED_VALUE
+				## Indent min
+				G_WHIP_DEFAULT_ITEM=$BW_INDENT_MIN
+				G_WHIP_INPUTBOX 'Please set the minimum offset' && BW_INDENT_MIN=$G_WHIP_RETURNED_VALUE
```
We should probably skip this if "min" was not selected? I know it can still be relevant if the line is still too long after colon-alignment, but the default 3 (equals dash) alignment should be fine in that case. I am just thinking that two additional prompts on each menu call is a bit much.

mtekman replied:
agreed

MichaIng commented on dietpi/func/dietpi-banner:
```diff
@@ -327,19 +350,35 @@ $GREEN_LINE"
+					'min' 'Indent to a fixed minimum offset'
+					'dash' 'Indent to the green dash character'
+					'colon' 'Indent to the green colon character'
+					'dashcolon' 'Indent first to Colon then Dash'
```
This does not seem to work yet. It either aligns to colon, the last one in the `make_whitespace_offset` function, or to `INDENT_MIN`, the fallback in main awk block. But checking this in awk script.

EDIT: Ah sorry. So this is for lines which do no have a colon. I guess we could merge this option with `colon`. I don't think anyone would want to have these not aligned at all (or to `INDENT_MIN` directly) if alignment to dash would work well? So we had:

- `min` = fixed indent, probably we should name it `fixed`?
- `dash` = intend everything to green dash, and `INDENT_MIN` as fallback
- `colon` = intend everything to colon where one exists, otherwise to dash, and `INDENT_MIN` as fallback

And maybe `INDENT_MIN` should be a fixed `1` if `fixed` option was not selected, otherwise its default does not make dash-aligned lines shorter, and depending on user input it might even make the line longer, where it was intended to make it shorter. So basically `INDENT_MIN` variable becomes `INDENT_FIXED`, and where it was used as fallback for the other options, we use a fixed `1`?

mtekman replied:
oh yep, I could have been a bit clearer with the definitions there.

`INDENT_MIN` was meant as the absolute fallback indentation strategy if no colon or dash was discovered in the line. Some users might want this min to be wider than the dash indent, but yes probably most will not.

> So basically INDENT_MIN variable becomes INDENT_FIXED, and where it was used as fallback for the other options, we use a fixed 1?

Sounds good!

MichaIng committed:
dietpi-banner: enhance mawk workaround

It cannot deal with the special line char with wildcard directly, but wrapping it into square bracket works.

MichaIng commented on dietpi/func/dietpi-banner-wrap.awk:
```diff
@@ -0,0 +1,96 @@
+    }
+    ## Green Lines: Truncate to MAXCOL
+    if (match($0, /────/)){
+        new_grn_line = sprintf("%*s", MAXCOL - 2, "") ## new line: of spaces
```
Maybe we should cap it at 54, which is the current right side column if word-wrapping is not enabled? So the lines do not change on larger terminals, but only if smaller than 54 columns.

`MAXCOL` itself of course needs to remain `tput cols`, to hide the ASCII art hostname and wrap other larger output lines correctly. Just for these green separator lines the cap would make sense.

mtekman replied:
sounds good to me!

MichaIng committed:
dietpi-banner: fix awk script name

MichaIng committed:
changelog: add entry for dietpi-banner word-wrap

MichaIng reviewed (approved):
I left some comments to discuss, but will merge this now for the first beta. Many thanks, works like a charm with the awk script and not affecting performance if not enabled, now that I replaced `cat` with a direct output by splitting it out into a dedicated function.

After some more testing and fine-tuning, I guess we can enable it by default. Will do some measurements, but I think the performance impact is negligible.
